pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+
+        self._trainable_params = set()
+        self._graph = None
+        self._output_dim = 0
+
+        self.wires = qml.wires.Wires([])
+        self.num_wires = 0
+
+        self.jacobian_options = {}
+
+        self.hash = 0
+        self.is_sampled = False
+
+        self._stack = None
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: wires={self.wires}, params={self.num_params}>"
+
+    def __enter__(self):
+        if not QueuingContext.recording():
+            # if the tape is the first active queuing context
+            # monkeypatch the operations to support the new queuing context
+            with contextlib.ExitStack() as stack:
+                for mock in mock_operations():
+                    stack.enter_context(mock)
+                self._stack = stack.pop_all()
+
+        QueuingContext.append(self)
+        return super().__enter__()
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        super().__exit__(exception_type, exception_value, traceback)
+
+        if not QueuingContext.recording():
+            # remove the monkeypatching
+            self._stack.__exit__(exception_type, exception_value, traceback)
+
+        self._process_queue()
+
+    @property
+    def interface(self):
+        """str, None: automatic differentiation interface used by the quantum tap (if any)"""
+        return None
+
+    # ========================================================
+    # construction methods
+    # ========================================================
+
+    def _process_queue(self):
+        """Process the annotated queue, creating a list of quantum
+        operations and measurement processes.
+
+        This method sets the following attributes:
+
+        * ``_ops``
+        * ``_obs``
+        * ``_par_info``
+        * ``_output_dim``
+        * ``_trainable_params``
+        * ``is_sampled``
+        """
+        self._prep = []
+        self._ops = []
+        self._obs = []
+        self._output_dim = 0
+
+        for obj, info in self._queue.items():
+
+            if isinstance(obj, QuantumTape):
+                self._ops.append(obj)
+
+            elif isinstance(obj, qml.operation.Operation) and not info.get("owner", False):
+                # operation objects with no owners
+                if self._obs:
+                    raise ValueError(
+                        f"Quantum operation {obj} must occur prior to any measurements."
+                    )
+
+                obj.do_check_domain = False
+
+                # invert the operation if required
+                obj.inverse = info.get("inverse", False)
+
+                if isinstance(obj, STATE_PREP_OPS):
+                    if self._ops:
+                        raise ValueError(
+                            f"State preperation operation {obj} must occur prior to any quantum operations."
+                        )
+
+                    self._prep.append(obj)
```

Comment:
Yep, this was deliberate to an extent, I didn't want to make any decisions wrt state prep validation in this PR (thought it would make sense to implement it in a future PR, when we also have a way of marking state prep operations).